crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -538,8 +457,43 @@ where
                 // to this event
                 let receiver = msg.message.packet_data.receiver.to_string();
                 let addr = TAddrData::Ibc(receiver.clone());
-                acc.decoder.insert(ibc_taddr(receiver), addr);
-                acc = self.apply_send_packet(acc, msg, keys_changed)?;
+                acc.decoder.insert(ibc_taddr(receiver.clone()), addr);
+                let ibc_trace = msg.message.packet_data.token.denom.to_string();
+                let amount = msg
+                    .message
+                    .packet_data
+                    .token
+                    .amount
+                    .try_into()
+                    .into_storage_result()?;
+                acc = self.apply_send_packet(
+                    acc,
+                    &msg.message.port_id_on_a,
+                    &msg.message.chan_id_on_a,
+                    ibc_trace,
+                    amount,
+                    receiver,
+                )?;
+            }
+            IbcMessage::NftTransfer(msg) => {
+                let receiver = msg.message.packet_data.receiver.to_string();
+                let addr = TAddrData::Ibc(receiver.clone());
+                acc.decoder.insert(ibc_taddr(receiver.clone()), addr);
+                for token_id in &msg.message.packet_data.token_ids.0 {
+                    let ibc_trace = format!(
+                        "{}/{}",
+                        msg.message.packet_data.class_id, token_id
+                    );
+                    let amount = Amount::from_u64(1);
+                    acc = self.apply_send_packet(
+                        acc,
+                        &msg.message.port_id_on_a,
+                        &msg.message.chan_id_on_a,
+                        ibc_trace,
+                        amount,
+                        &receiver,
+                    )?;
+                }
             }
             // This event is emitted on the receiver
             IbcMessage::RecvPacket(msg) => {
```

Comment:
Below we try to decode `msg.message.packet.data` as `namada_core::ibc::apps::transfer::types::packet::PacketData`. If we want to support NFT, would we also need to try and decode the data as a `namada_core::ibc::apps::nft_transfer::types::packet::PacketData`? Or maybe branch the execution based on `msg.message.packet.port_id_on_a` or `msg.message.packet.port_id_on_b`?

Reply:
yes, I'll fix it for `apply_write_ack`